### ISSUE TITLE
Question: injection in custom loader at parsing stage

### ISSUE THREAD

**Issue by exteon** [goaop/framework#431]
Hi,
This is an usage question, not sure if this is the right place to ask but anyway:
We have a framework that implements custom loaders that replicate 50% of your workflow, in the sense that it too rewrites source code to accomplish some features such as drop-in class-replacement -or-chaining plugins, then feeds it through a stream wrapper to mask the changes and feed the original code to the debugger.
We'd like to use Go!AOP but in order to do that, we need to plug into the Go!AOP process with the already dynamically modified/proxied sources. So we would need to basically skip the file-loading process and just trigger "here's the < class >, here's the < proxied_source >, do your magic" then either receive back the patched file reference, or at this point we could just defer the actual loading to Go!AOP. We would prefer a reference here, because we also implement a partial modification watcher and partial cache-invalidation for the classes, so if we defer to Go!AOP we would also basically need to tap into Go!AOP-s caching process... Which I haven't looked into yet but expect to be a bit tricky. What would be the best way to go around this?

**Comment by lisachenko** [comment]
Hi, @exteon!

From your description it's unclear how to integrate your system with Go! AOP engine. Because class loading in framework really tricky to work with stream wrappers and opcache. You can give only one file to AOP engine to transform/load and can't control another proxied source files...

**Comment by exteon** [comment]
Well, I figured it should work more or less similar to the Doctrine loaders, but I haven't been able to figure that out precisely how it works. But the concept is sort-of the same: Doctrine also takes an initial set of class files/defs and creates a first layer of proxy around them, then Go! AOP adds the second layer of proxying... am I correct so far?

The only difference we would like to make would be that, if possible, Go! AOP does not do the actual loading of the classes (no stream wrapping, no loading) but we get a reference to the created wrapper, i.e. maybe a path to `cache/path/to/class.php`, I think is feasible as Go! AOP seems to generate at least 2 files / proxied class; or maybe an array like 
```php
[
    'Class_AopProxied'=>'...Class_AopProxied code...',
    'Class'=>'...Class code...'
]
```
This we can feed into our own loading/caching/stream-wrapping system.

So basically what we'd like to do is, instead of Go! AOP hijacking the loading process, is calling Go! AOP explicitly from our own loader to get a class proxied.

**Comment by lisachenko** [comment]
I'm curious if you have succeeded with integration or not?

**Comment by exteon** [comment]
No, it's pending more research for now, as I couldn't find a straight way to do it on the first pass. I registered https://bugs.xdebug.org/view.php?id=1702 to see if we can also simplify the caching process, and, if you have any more insight into how to do this, it would really help.

**Comment by lisachenko** [comment]
Closed by inactivity. Ping me if something is needed

**Comment by exteon** [comment]
Hi Alexander,
Just wanted to follow up one year later, I published https://github.com/exteon/mapping-class-loader ; I plan to make into a modular library to go to for advanced loader functionalities; I put aside the AOP issues for now, but I just wanted to let you know and ask if you see any use for it.
